### DISCUSSION TITLE
The GUI

### DISCUSSION THREAD

**Discussion by grd** [grd/FreePDM#77] [General]
I have spent way too much time in the FileSystem, simply because it is such a huge part of the PDM. But now it is about time to finish it.

What is next? The GUI. I am not a GUI expert and I probably am making tons of stupid mistakes but it needs to be here.

About what GUI? Well, I am not a fan of Python (PySide is pretty okay, but it looks outdated) and I like Nim a lot more, but I realized soon that compiling Qt takes a lot of time with Nim. Too much (about half a minute on my laptop). That is why I went searching for alternatives. BTW, they all have a smell. And if there is one PL that I like even more than Nim it is Go, because it is so simple and I am just a simple programmer.

That is why I looked for a multi-platform GUI with Go, maybe also with Android and WASM. There are only two that can do that. GIO and Fyne. GIO is probably the best choice if you care about performance and tweaking. I care about neither. Fyne is probably the best choice if you just want to get things done. 

Two benefits of Go is that:
1. I can use all in one place and even reuse some code.
2. I can compile the GUI into one massive executable file, so I don't have any problems with AppImage and the like.

That is it. If you have any suggestions or ideas, please put a note here.

I think that I am going to start at the end of this month.

Edit: 
About Fyne. Have a look here. https://www.youtube.com/watch?v=EsX8CgC_N5M&t=565s 
It looks miles ahead of Qt Widgets. QML is probably nicer but then you are into the land of magic and I don't like that.

About the communication between the server and the client, I will be using ~~gRPC~~ JSON, that's a simple protocol. There are also ways to use ~~gRPC~~ JSON with Python. So maybe we can also make a Python client.

**Comment by JettisonableMass** [comment]
@grd I'm commenting on this discussion as I wasn't sure how else to reach out! I am interested in the gui though! I'm really excited to see this project develop. I have a moderate amount experience with managed CAD systems professionally, as a designer and a rookie administrator. Would be interested to learn more and help if possible / useful!

**Comment by grd** [comment]
Right now I am working on a Fyne GUI, with the name fpg (FreePDM GUI). It is still in its infancy state, but I am working on the vaults. It is supposed to work with rsync so that we don´t have the problems with SMB anymore. Which me luck ;-)
